test/evp_extra_test.c

```diff
@@ -1833,7 +1959,133 @@ static int test_decrypt_null_chunks(void)
 }
 #endif /* !defined(OPENSSL_NO_CHACHA) && !defined(OPENSSL_NO_POLY1305) */
 
-#if !defined(OPENSSL_NO_DH) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+#ifndef OPENSSL_NO_DH
+/* Test combinations of private, public, missing and private + public key 
+ * params to ensure they are all accepted */
+static int test_DH_priv_pub(void)
+{
+    OSSL_PARAM_BLD *bld = NULL;
+    OSSL_PARAM *params = NULL;
+    BIGNUM *p = NULL, *q = NULL, *g = NULL, *pub = NULL, *priv = NULL;
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    int ret = 0;
+
+    /*
+     * Setup the parameters for our DH object. For our purposes they don't
+     * have to actually be *valid* parameters. We just need to set something.
+     */
+    if (!TEST_ptr(pctx = EVP_PKEY_CTX_new_from_name(testctx, "DH", NULL))
```

Comment:
Does this function and the DSA test above differ more than the key type string?

Reply:
Yes I believe so. Wasn't sure if the behaviour would be the same when I copy/pasted it but it does seem to be. I should probably refactor it...

Reply:
I would appreciate a merge of the two functions...  I firmly dislike code copying en masse, perhaps it's just me...

Reply:
I'm assuming a model with a common function, and then individual test functions that call it with different parameters.

Reply:
Should be fixed in the new commit :)